usr/lib/common/pin_get.c

```diff
@@ -0,0 +1,136 @@
+/*
+ * COPYRIGHT (c) International Business Machines Corp. 2022
+ *
+ * This program is provided under the terms of the Common Public License,
+ * version 1.0 (CPL-1.0). Any use, reproduction or distribution for this
+ * software constitutes recipient's acceptance of CPL-1.0 terms which can be
+ * found in the file LICENSE file or at
+ * https://opensource.org/licenses/cpl1.0.php
+ */
+
+#define _POSIX_C_SOURCE 200809L
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <termios.h>
+#include <unistd.h>
+#include <sys/mman.h>
+
+#include <pin_get.h>
+
+struct pinbuf pinbuf_empty = {
+    .p = NULL,
+    .len = 0,
+};
+
+static void echo(bool on)
+{
+    struct termios term;
+
+    tcgetattr(fileno(stdin), &term);
+
+    if (on)
+        term.c_lflag |= ECHO;
+    else
+        term.c_lflag &= ~ECHO;
+
+    tcsetattr(fileno(stdin), TCSAFLUSH, &term);
+}
+
+const char *pin_prompt(struct pinbuf *buf, const char *msg)
+{
+    ssize_t n;
+
+    if (!buf)
+        return NULL;
+
+    printf("%s", msg);
+    fflush(stdout);
+    echo(false);
+
+    n = getline(&buf->p, &buf->len, stdin);
+
+    echo(true);
+    printf("\n");
+    fflush(stdout);
+
+    /* delayed getline() error handling */
+    if (n == -1)
+        return NULL;
+
+    /* strip CR/LF (if any) */
+    (buf->p)[strcspn(buf->p, "\r\n")] = '\0';
+
+    buf->len = strlen(buf->p);
+    return buf->p;
+}
+
+const char *pin_get_new(struct pinbuf *buf, const char *msg1, const char *msg2)
+{
+    const char *pin = NULL;
+    struct pinbuf buf2 = pinbuf_empty;
+
+    if (!pin_prompt(buf, msg1) ||
+        !pin_prompt(&buf2, msg2))
+        return NULL;
+
+    if ((buf->len && buf2.len) &&
+        (buf->len == buf2.len) &&
+        (!strncmp(buf->p, buf2.p, buf->len)))
```

Comment:
if you are about to allow \0 within the pin, then you would need to use memcmp here.

Reply:
The piece of code is wrong. the len variables do not represent the allocated space in p, not the length of the strings. I'll fix this in the next version.